File: lib/src/message_codec.dart
import 'package:amap_flutter/amap_flutter.dart';
import 'package:flutter/services.dart';

class AMapApiCodec extends StandardMessageCodec {
  const AMapApiCodec();

  @override
  void writeValue(WriteBuffer buffer, Object? value) {
    if (value is MapType) {
      buffer.putUint8(128);
      writeValue(buffer, value.index);
    } else if (value is UIControlAnchor) {
      buffer.putUint8(129);
      writeValue(buffer, value.index);
    } else if (value is UserLocationType) {
      buffer.putUint8(130);
      writeValue(buffer, value.index);
    } else if (value is Anchor) {
      buffer.putUint8(131);
      writeValue(buffer, value.encode());
    } else if (value is Bitmap) {
      buffer.putUint8(132);
      writeValue(buffer, value.encode());
    } else if (value is CameraPosition) {
      buffer.putUint8(133);
      writeValue(buffer, value.encode());
    } else if (value is EdgePadding) {
      buffer.putUint8(134);
      writeValue(buffer, value.encode());
    } else if (value is Location) {
      buffer.putUint8(135);
      writeValue(buffer, value.encode());
    } else if (value is MapInitConfig) {
      buffer.putUint8(136);
      writeValue(buffer, value.encode());
    } else if (value is MapUpdateConfig) {
      buffer.putUint8(137);
      writeValue(buffer, value.encode());
    } else if (value is Marker) {
      buffer.putUint8(138);
      writeValue(buffer, value.encode());
    } else if (value is Poi) {
      buffer.putUint8(139);
      writeValue(buffer, value.encode());
    } else if (value is Position) {
      buffer.putUint8(140);
      writeValue(buffer, value.encode());
    } else if (value is Region) {
      buffer.putUint8(141);
      writeValue(buffer, value.encode());
    } else if (value is Size) {
      buffer.putUint8(142);
      writeValue(buffer, value.encode());
    } else if (value is UIControlOffset) {
      buffer.putUint8(143);
      writeValue(buffer, value.encode());
    } else if (value is UIControlPosition) {
      buffer.putUint8(144);
      writeValue(buffer, value.encode());
    } else if (value is UserLocationConfig) {
      buffer.putUint8(145);
      writeValue(buffer, value.encode());
    } else if (value is UserLocationStyle) {
      buffer.putUint8(146);
      writeValue(buffer, value.encode());
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  Object? readValueOfType(int type, ReadBuffer buffer) {
    switch (type) {
      case 128:
        return MapType.values.elementAt(readValue(buffer) as int);
      case 129:
        return UIControlAnchor.values.elementAt(readValue(buffer) as int);
      case 130:
        return UserLocationType.values.elementAt(readValue(buffer) as int);
      case 131:
        return Anchor.decode(readValue(buffer) as List<Object?>);
      case 132:
        return Bitmap.decode(readValue(buffer) as List<Object?>);
      case 133:
        return CameraPosition.decode(readValue(buffer) as List<Object?>);
      case 134:
        return EdgePadding.decode(readValue(buffer) as List<Object?>);
      case 135:
        return Location.decode(readValue(buffer) as List<Object?>);
      case 136:
        return MapInitConfig.decode(readValue(buffer) as List<Object?>);
      case 137:
        return MapUpdateConfig.decode(readValue(buffer) as List<Object?>);
      case 138:
        return Marker.decode(readValue(buffer) as List<Object?>);
      case 139:
        return Poi.decode(readValue(buffer) as List<Object?>);
      case 140:
        return Position.decode(readValue(buffer) as List<Object?>);
      case 141:
        return Region.decode(readValue(buffer) as List<Object?>);
      case 142:
        return Size.decode(readValue(buffer) as List<Object?>);
      case 143:
        return UIControlOffset.decode(readValue(buffer) as List<Object?>);
      case 144:
        return UIControlPosition.decode(readValue(buffer) as List<Object?>);
      case 145:
        return UserLocationConfig.decode(readValue(buffer) as List<Object?>);
      case 146:
        return UserLocationStyle.decode(readValue(buffer) as List<Object?>);
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}

File: android/src/main/kotlin/com/morbit/amap_flutter/MessageCodec.kt
package com.morbit.amap_flutter

import io.flutter.plugin.common.StandardMessageCodec
import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer

object AMapApiCodec : StandardMessageCodec() {
  override fun readValueOfType(type: Byte, buffer: ByteBuffer): Any? {
    return when (type) {
      128.toByte() -> {
        return (readValue(buffer) as? Int)?.let {
          MapType.ofRaw(it)
        }
      }

      129.toByte() -> {
        return (readValue(buffer) as? Int)?.let {
          UIControlAnchor.ofRaw(it)
        }
      }

      130.toByte() -> {
        return (readValue(buffer) as? Int)?.let {
          UserLocationType.ofRaw(it)
        }
      }

      131.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Anchor.fromList(it)
        }
      }

      132.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Bitmap.fromList(it)
        }
      }

      133.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          CameraPosition.fromList(it)
        }
      }

      134.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          EdgePadding.fromList(it)
        }
      }

      135.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Location.fromList(it)
        }
      }

      136.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          MapInitConfig.fromList(it)
        }
      }

      137.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          MapUpdateConfig.fromList(it)
        }
      }

      138.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Marker.fromList(it)
        }
      }

      139.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Poi.fromList(it)
        }
      }

      140.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Position.fromList(it)
        }
      }

      141.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Region.fromList(it)
        }
      }

      142.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          Size.fromList(it)
        }
      }

      143.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          UIControlOffset.fromList(it)
        }
      }

      144.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          UIControlPosition.fromList(it)
        }
      }

      145.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          UserLocationConfig.fromList(it)
        }
      }

      146.toByte() -> {
        return (readValue(buffer) as? List<Any?>)?.let {
          UserLocationStyle.fromList(it)
        }
      }

      else -> super.readValueOfType(type, buffer)
    }
  }

  override fun writeValue(stream: ByteArrayOutputStream, value: Any?) {
    when (value) {
      is MapType -> {
        stream.write(128)
        writeValue(stream, value.raw)
      }

      is UIControlAnchor -> {
        stream.write(129)
        writeValue(stream, value.raw)
      }

      is UserLocationType -> {
        stream.write(130)
        writeValue(stream, value.raw)
      }

      is Anchor -> {
        stream.write(131)
        writeValue(stream, value.toList())
      }

      is Bitmap -> {
        stream.write(132)
        writeValue(stream, value.toList())
      }

      is CameraPosition -> {
        stream.write(133)
        writeValue(stream, value.toList())
      }

      is EdgePadding -> {
        stream.write(134)
        writeValue(stream, value.toList())
      }

      is Location -> {
        stream.write(135)
        writeValue(stream, value.toList())
      }

      is MapInitConfig -> {
        stream.write(136)
        writeValue(stream, value.toList())
      }

      is MapUpdateConfig -> {
        stream.write(137)
        writeValue(stream, value.toList())
      }

      is Marker -> {
        stream.write(138)
        writeValue(stream, value.toList())
      }

      is Poi -> {
        stream.write(139)
        writeValue(stream, value.toList())
      }

      is Position -> {
        stream.write(140)
        writeValue(stream, value.toList())
      }

      is Region -> {
        stream.write(141)
        writeValue(stream, value.toList())
      }

      is Size -> {
        stream.write(142)
        writeValue(stream, value.toList())
      }

      is UIControlOffset -> {
        stream.write(143)
        writeValue(stream, value.toList())
      }

      is UIControlPosition -> {
        stream.write(144)
        writeValue(stream, value.toList())
      }

      is UserLocationConfig -> {
        stream.write(145)
        writeValue(stream, value.toList())
      }

      is UserLocationStyle -> {
        stream.write(146)
        writeValue(stream, value.toList())
      }

      else -> super.writeValue(stream, value)
    }
  }
}